Tasks/DownloadBuildArtifacts/main.ts

```diff
@@ -0,0 +1,151 @@
+var path = require('path')
+var url = require('url')
+
+import * as tl from 'vsts-task-lib/task';
+import { IBuildApi } from 'vso-node-api/BuildApi';
+import { IRequestHandler } from 'vso-node-api/interfaces/common/VsoBaseInterfaces';
+import { WebApi, getHandlerFromToken } from 'vso-node-api/WebApi';
+
+import * as models from 'item-level-downloader/Models';
+import * as engine from 'item-level-downloader/Engine';
+import * as providers from 'item-level-downloader/Providers';
+import * as webHandlers from 'item-level-downloader/Providers/Handlers';
+
+tl.setResourcePath(path.join(__dirname, 'task.json'));
+
+async function main(): Promise<void> {
+    var promise = new Promise<void>(async (resolve, reject) => {
+        var buildType: string = tl.getInput("buildType", true);
+        var isCurrentBuild: boolean = buildType.toLowerCase() === 'current';
+        var projectId: string = isCurrentBuild ? tl.getVariable("System.TeamProjectId") : tl.getInput("project", true);
+        var definitionId: string = isCurrentBuild ? '' : tl.getInput("definition", true);
+        var buildId: number = parseInt(isCurrentBuild ? tl.getVariable("Build.BuildId") : tl.getInput("buildId", true));
+        var downloadPath: string = tl.getInput("downloadPath", true);
+        var downloadType: string = tl.getInput("downloadType", true);
+
+        var endpointUrl: string = tl.getVariable("System.TeamFoundationCollectionUri");
+        var accessToken: string = tl.getEndpointAuthorizationParameter('SYSTEMVSSCONNECTION', 'AccessToken', false);
+        var credentialHandler: IRequestHandler = getHandlerFromToken(accessToken);
+        var webApi: WebApi = new WebApi(endpointUrl, credentialHandler);
+        var debugMode: string = tl.getVariable('System.Debug');
+        var isVerbose: boolean = debugMode ? debugMode.toLowerCase() != 'false' : false;
+        var parallelLimit: number = +tl.getVariable("release.artifact.download.parallellimit");
+
+        var templatePath: string = path.join(__dirname, 'vsts.handlebars.txt');
+        var buildApi: IBuildApi = webApi.getBuildApi();
+        var artifacts = [];
+        var itemPattern: string = '**';
+
+        // verfiy that buildId belongs to the definition selected
+        if (definitionId) {
```

Comment:
What is the point of the definition id check?

Reply:
@scottdallamura - this is just a safety net, since build definition id was given as task input. in cases where the build no. doesn't appear in drop-down (we have a limit on no. of builds we query in drop-down), user can type in build id in the input in which case this validation will be of use (to make sure user didn't get the id wrong accidentally).

Reply:
I see.  This sounds like a hack to cover up for a flawed experience elsewhere.  Does it add any real value?  In general, we don't add checks far downstream to cover up for bugs earlier - we should fix the initial issue.

Reply:
We cannot get away from having build definition as task input as it is required for a good task authoring experience. 

If task doesn't validate, where else can we catch user giving mismatched build definition id & build id?